lib/NMISNG/WMI.pm

```diff
@@ -181,163 +183,214 @@ sub _run_query
 	my (@rawdata, $exitcode, %result, $version, $cmdLine);
 	my $v2option = "";
 
-	# fork and pipe
-	my $pid = open(WMIC, "-|");
-	if (!defined $pid)
+	if (!$self->{version})
 	{
-		unlink $cmdfn or warn "Could not remove file '$cmdfn' Error: $!";
-		unlink $tfn or warn "Could not remove file '$tfn' Error: $!";
-		unlink $authfn or warn "Could not remove file '$authfn' Error: $!";
-		return (error => "cannot fork to run wmic: $!");
+		$version = "Version 1";
 	}
-	elsif ($pid)
-	{
-		# parent: save and restore any previously running alarm,
-		# but don't bother subtracting time spent here
-		my $remaining = alarm(0);
-		eval
-		{
-			local $SIG{ALRM} = sub { die "alarm\n"; };
-			alarm($timeout) if ($timeout); # setup execution timeout
 
-			close $tfh;									# not ours to use
-			close $authfh;
-			@rawdata = <WMIC>;					# read the goodies from the child
-			close(WMIC);
-			$exitcode = $?;
-			alarm(0);
-		};
-		alarm($remaining) if ($remaining);
-		if ($@ and $@ eq "alarm\n")
-		{
-			# don't want the wmic process to hang around, we stopped consuming its output
-			# and it can't do anything useful anymore
-			kill("KILL",$pid);
-			unlink $cmdfn or warn "Could not remove file '$cmdfn' Error: $!";
-			unlink $tfn or warn "Could not remove file '$tfn' Error: $!";
-			unlink $authfn or warn "Could not remove file '$authfn' Error: $!";
-			return (error => "timeout after $timeout seconds");
-		}
-	}
-	else
+	if( $self->{version} eq 'wmic_server' ) 
 	{
-		# child
-		open(STDIN, "</dev/null");
-		open(STDERR, '>&'.$tfh);		# stderr to go there, please
-
-		# -A format is badly documented. smbclient manpage has a little bit of info
-		# however, unclear if that password can be quoted or contain spaces or the like...
+		require Mojo::UserAgent;
+		my $headers = {"Content-type" => 'application/json', Accept => 'application/json'};
+		my $client  = Mojo::UserAgent->new();
+
+		my $url     = Mojo::URL->new( $self->{wmic_server_location} );
+		my $id = $self->{username};
+		my $token = 'MYSECRETUSERACCESSTOKEN1'; # // $self->{password};
+		my $host = $self->{host};
+		my $namespace = "root/cimv2";
+		my $post_json = { id => $id, token => $token, host => $host, query => $query, namespace => $namespace };
+
+		my $res;
+		# connection errors die so the request needs to be wrapped in a try
+		try {
+			$res = $client->post($url => $headers => json => $post_json )->result;
```

Comment:
do we not want to use the Mojo::UserAgent error handling?

Reply:
We do further down.
From the docs: # Fine grained response handling (dies on connection errors), so without a try or eval the whole worker dies an things go bad for the whole system